### ISSUE TITLE
Not populated after create, drop

### ISSUE THREAD

**Issue by stuartdrennan** [attr-encrypted/attr_encrypted#104]
Hi all

Seem to have a strange issue, might be something silly I'm doing. 

If I rake db:drop then rake db:create the database manually, and then these tasks are ran again in a rake script, the encrypted fields populated as part of this same rake task don't get populated, but no error is returned. If I run this rake task (without the first set of drop create) the fields populate fine.

Any thoughts?

Thank, Stuart

**Comment by monfresh** [comment]
I am experiencing a similar issue as well. In a migration, I have the following block:
```ruby
SecurityAnswer.find_each do |answer|
  answer.text = answer.body
  answer.save!
end
```
After I run the migration, when I run this in the console:
```
rails c
SecurityAnswer.pluck(:encrypted_text)
```
I get `nil` everywhere.

However, if I run the original block again, it populates the `encrypted_text` field properly this time. It's only during the migration that it doesn't do it. It's very strange!

I also added some `puts` statements in the migration, and I can see that `answer.text` has an encrypted value. But somehow, after `rake db:migrate` is done, it gets set to nil.

**Comment by monfresh** [comment]
Also, I should mention that I have also tried separating the migrations, and resetting the column information, as discussed in #59, but it's still not working. Here are my migrations:
```ruby
def change
  change_table :security_answers do |t|
    t.string :encrypted_text
    t.string :encrypted_text_iv
    t.string :encrypted_text_salt
  end
end
```
and
```ruby
def up
  SecurityAnswer.reset_column_information
  SecurityAnswer.find_each do |answer|
    answer.text = answer.body
    answer.save!
  end
end
```

**Comment by monfresh** [comment]
So, the problem is when running both migrations in the same run with `rake db:migrate`. If you run the first migration alone with `rake db:migrate`, then add the second migration and run `rake db:migrate`, it works.

**Comment by monfresh** [comment]
Just to clarify, my comment above was meant to provide steps to reproduce. I believe this is a bug. You should be able to run multiple migrations in the same rake task.

**Comment by saghaulor** [comment]
@monfresh I have encountered the same issue. When I encountered this issue I assumed that it was relative to my code base. Unfortunately, I have not been able to determine the cause. There are two work arounds presently, run the migrations in stages, or write the encryption migration such that it doesn't use the accessors that attr_encrypted would give you. However, this means that you're effectively rewriting the accessors. If you are able to find out more information it would be extremely helpful.

**Comment by afeld** [comment]
I was debugging this w/ @monfresh... my guess is that there is some configuration or something loaded on app start-up (a.k.a. when the files are read and the classes are defined) which is not being reloaded between when the columns are created and when they are being accessed. Need to look into it, though.

**Comment by duffyjp** [comment]
I believe I'm experiencing the same issue.  Running a spec which creates a person, then sets an ssn, then saves.  If you delete the test DB and run rspec, AR makes to attempt to save encrypted_ssn on the person.  Run it a second time after the test DB is already there and it does.

Here is the interesting part of a diff between the two test logs.
![screen shot 2016-01-28 at 9 42 04 am](https://cloud.githubusercontent.com/assets/382216/12649523/8c3f417e-c5a4-11e5-84e5-99954ba38a73.png)

Notice the missing ```UPDATE "people" SET "encrypted_ssn" ...``` on the first run.

```ruby
it "should save the ssn to encrypted_ssn" do
  person = create :person
  person.ssn = "123456789"
  person.save!  # Doesn't actually write to DB the first pass
  expect(person.reload.ssn).to eq("123456789")  # Always successful, oddly.
  expect(Person.find(person.id).ssn).to eq "123456789"  # Fails the first pass, passes the second
end
```

**Comment by peterox** [comment]
Has anyone tracked down the cause of this? I am also seeing this behaviour.

Has v2 fixed this?

**Comment by monfresh** [comment]
I tracked down the cause of this a while back and have been meaning to update this issue, but haven't had time. The bug is definitely in this gem, and if I recall correctly, it's due to the metaprogramming that defines and makes use of the getter and setter methods corresponding to the database columns. 

I think the problem is that Rails loads the application code only once before it starts the migration, and when the migration gets to the point where it needs to use the setter method (to set the encrypted value), it doesn't work because Rails doesn't know about this new setter method because when the `attr_encrypted` gem was first loaded, the new encrypted DB column had not been created yet, and therefore the gem had not created the setter method dynamically yet.

If the methods exist beforehand, then it works properly. For example, I needed to add encryption to a particular column in the `two_factor_authentication` gem, and to do that, instead of using `attr_encrypted` as is, I extracted the portions I needed and wrote the methods directly. You can see the code here: https://github.com/Houdini/two_factor_authentication/blob/master/lib/two_factor_authentication/models/two_factor_authenticatable.rb#L66-L139

**Comment by saghaulor** [comment]
@peterox v2 has not fixed this. 

@monfresh I believe this isn't a problem of the gem per se. This is a general problem of Rails and db migrations. Manipulating data in migrations using your model is known risky behavior. Justin Searls has a blog post that covers why it's a bad idea. http://blog.testdouble.com/posts/2014-11-04-healthy-migration-habits.html#habit-4-don-t-reference-models

In fact, this is such a bad pattern that Tenderlove even wrote a gist that disables model loading in migrations. Then Searls wrote a gem that essentially encompassed that behavior. http://blog.testdouble.com/posts/2016-02-05-good-migrations.html

We may be able to do something in the gem to help with this issue, but I think the best advice is to not manipulate using Rails models in your migrations.

**Comment by BrunoRomes** [comment]
Has any solution been found? I'm experiencing this problem too. My scenario is just like the one @duffyjp described.
This problem is breaking my CI environment, since I always drop and recreate the test database.

**Comment by tobmatth** [comment]
The problem seems to be this Line: https://github.com/attr-encrypted/attr_encrypted/blob/master/lib/attr_encrypted.rb#L143 (and the following lines for salt and iv) which checks if the Model defines an accessor for encrypted_attribute_name and creates an accessor in case it doesn't. However, adding columns doesn't seem to update a Models instance methods once the Class is loaded, which in this case causes an accessor to be created for an existing database column, thus preventing the data to be persisted.

One option would be make the check adapter specific, e.g. use `Class.column_names.include? attr_name`, but i'm not sure it's worth it. 

In short:
```ruby
> SystemSetting.instance_methods.include? :test
=> false
> ActiveRecord::Migration.add_column :system_settings, :test, :string
=> #<PG::Result:0xb2e7818 status=PGRES_COMMAND_OK ntuples=0 nfields=0 cmd_tuples=0>
> SystemSetting.reset_column_information
=> nil
> SystemSetting.instance_methods.include? :test
=> false
> SystemSetting.column_names.include? 'test'
=> true
```

@saghaulor, @BrunoRomes Any thoughts on this?

**Comment by saghaulor** [comment]
@tobmatth After adding the column and restarting the app does the problem persist? It has been my experience that ```:reset_column_information``` isn't enough.

**Comment by tobmatth** [comment]
The Problem is gone as soon as Classes are reloaded, or a new instance of the affected Model is built. `:reset_column_information` doesn't help here.

**Comment by tobmatth** [comment]
@saghaulor Forget about my comment regarding an adapter specific check, i was misunderstanding the issue. The only solution to this is not to reference Models in Migrations, or to force class reloading...

**Comment by duffyjp** [comment]
@tobmatth 
> ...force class reloading

Has that worked for you?  What was your technique?  I've tried a number of things and it hasn't made any difference.  I'm working on a migration upgrading my attr_encrypted version and the only way I can do it is to migrate then run a rake task that does the actual data conversion (new _iv columns). 

I'd love to do this all in one shot.

**Comment by tobmatth** [comment]
@duffyjp try using

    ActionDispatch::Reloader.cleanup!
    ActionDispatch::Reloader.prepare!

after modifying columns.

HTH
Tobi

**Comment by duffyjp** [comment]
@tobmatth It worked!  Amazing-- That was the golden ticket. 

Thanks!

**Comment by BrunoChauvet** [comment]
`ActionDispatch::Reloader.prepare!` may cause issues with JRuby and raise `LoadError`.
The workaround is to unload and reload models this way when modifying schema structure such as migrating from 1.3 to 3.x:
```ruby
# Preserve existing encrypted attribute (v1.3)
rename_column :models, :encrypted_password, :encrypted_password_old
# Add encrypted attribute (v3.0)
add_column :models, :encrypted_password, :text
add_column :models, :encrypted_password_iv, :string

# Force reloading the model
Object.send(:remove_const, :Model)
ActiveRecord::Base.connection.schema_cache.clear!
load File.join(Rails.root, 'app', 'models', 'model.rb')

# Copy over encrypted values
Model.all.find_each { |m| m.update_attribute(encrypted_password: m.encrypted_password_old) }
```

**Comment by sapientpants** [comment]
```ruby
Rails.application.reloader.reload!
```

Works with Rails 5.2. I have no idea if this was available in earlier versions of Rails.
